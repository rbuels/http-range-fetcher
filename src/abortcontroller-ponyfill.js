/* eslint-disable */
if (typeof AbortController === 'undefined') {
  const {
    AbortController,
    AbortSignal,
  } = require('abortcontroller-polyfill/dist/cjs-ponyfill')
  module.exports = { AbortController, AbortSignal }
} else {
  module.exports = { AbortController, AbortSignal }
}
