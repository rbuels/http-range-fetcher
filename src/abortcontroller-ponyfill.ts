//@ts-nocheck
/* eslint-disable */

import {
  AbortController as ponyfillAbortController,
  AbortSignal as ponyfillAbortSignal,
} from 'abortcontroller-polyfill/dist/cjs-ponyfill'
var getGlobal = function () {
  // the only reliable means to get the global object is
  // `Function('return this')()`
  // However, this causes CSP violations in Chrome apps.
  if (typeof self !== 'undefined') {
    return self
  }
  if (typeof window !== 'undefined') {
    return window
  }
  if (typeof global !== 'undefined') {
    return global
  }
  throw new Error('unable to locate global object')
}

//@ts-ignore
let AbortController =
  typeof getGlobal().AbortController === 'undefined'
    ? ponyfillAbortController
    : getGlobal().AbortController
//@ts-ignore
let AbortSignal =
  typeof getGlobal().AbortController === 'undefined'
    ? ponyfillAbortSignal
    : getGlobal().AbortSignal

export { AbortController, AbortSignal }
