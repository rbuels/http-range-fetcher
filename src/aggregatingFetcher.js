/**
 * takes fetch requests and aggregates them at a certain time frequency
 */
class AggregatingFetcher {
  /**
   *
   * @param {object} params
   * @param {number} [params.frequency] number of milliseconds to wait for requests to aggregate
   */
  constructor({
    frequency = 100,
    fetch,
    maxExtraSize = 32000,
    maxFetchSize = 1000000,
  }) {
    this.requestQueues = {} // url => array of requests
    this.fetchCallback = fetch
    this.frequency = frequency
    this.maxExtraSize = maxExtraSize
    this.maxFetchSize = maxFetchSize
  }

  _canAggregate(requestGroup, request) {
    return (
      // the fetches overlap, or come close
      request.start <= requestGroup.end + this.maxExtraSize &&
      // aggregating would not result in a fetch that is too big
      request.end - request.start + requestGroup.end - requestGroup.start <
        this.maxFetchSize
    )
  }

  // dispatch a request group as a single request
  // and then slice the result back up to satisfy
  // the individual requests
  _dispatch({ url, start, end, requests }) {
    this.fetchCallback(url, start, end).then(
      data => {
        requests.forEach(({ start: reqStart, end: reqEnd, resolve }) => {
          // remember Buffer.slice does not copy, it creates
          // an offset child buffer pointing to the same data
          resolve(data.slice(reqStart - start, reqEnd - start))
        })
      },
      err => requests.forEach(({ reject }) => reject(err)),
    )
  }

  _aggregateAndDispatch() {
    Object.entries(this.requestQueues).forEach(([url, requests]) => {
      if (!requests || !requests.length) return
      // console.log(url, requests)
      // aggregate the requests in this url's queue
      const sortedRequests = requests.sort((a, b) => a[0] - b[0])
      let currentRequestGroup
      do {
        const next = sortedRequests.shift()
        if (
          currentRequestGroup &&
          this._canAggregate(currentRequestGroup, next)
        ) {
          // aggregate it
          currentRequestGroup.requests.push(next)
          currentRequestGroup.end = next.end
        } else {
          // out of range, dispatch the current request group
          if (currentRequestGroup) this._dispatch(currentRequestGroup)
          currentRequestGroup = {
            requests: [next],
            url,
            start: next.start,
            end: next.end,
          }
        }
      } while (sortedRequests.length)

      if (currentRequestGroup) this._dispatch(currentRequestGroup)
    })
  }

  _enQueue(url, request) {
    if (!this.requestQueues[url]) this.requestQueues[url] = []
    this.requestQueues[url].push(request)
  }

  /**
   *
   * @param {string} url
   * @param {number} start 0-based half-open
   * @param {number} end 0-based half-open
   */
  fetch(url, start, end) {
    return new Promise((resolve, reject) => {
      this._enQueue(url, { start, end, resolve, reject })
      if (!this.timeout) {
        this.timeout = setTimeout(() => {
          this.timeout = undefined
          this._aggregateAndDispatch()
        }, this.frequency)
      }
    })
  }
}

module.exports = AggregatingFetcher
