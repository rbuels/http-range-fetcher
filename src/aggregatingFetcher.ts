interface Result {
  headers: Headers
  buffer: Uint8Array
}

interface Req {
  start: number
  end: number
  reject: (arg: unknown) => void
  resolve: (arg: Result) => void
  requestOptions: { signal?: AbortSignal }
}

interface ReqGroup {
  requests: Req[]
  url: string
  start: number
  end: number
}

/**
 * takes fetch requests and aggregates them at a certain time frequency
 */
export default class AggregatingFetcher {
  timeout: unknown
  requestQueues: Record<string, Req[]>
  fetchCallback: (
    url: string,
    start: number,
    end: number,
    arg: { signal?: AbortSignal },
  ) => Promise<Result>
  frequency: number
  maxExtraSize: number
  maxFetchSize: number
  /**
   *
   * @param params
   * @param [params.frequency] number of milliseconds to wait for requests to aggregate
   */
  constructor({
    fetch,
    frequency = 100,
    maxExtraSize = 32000,
    maxFetchSize = 1000000,
  }: {
    fetch: (url: string, start: number, end: number) => Promise<Result>
    frequency: number
    maxExtraSize: number
    maxFetchSize: number
  }) {
    this.requestQueues = {} // url => array of requests
    this.fetchCallback = fetch
    this.frequency = frequency
    this.maxExtraSize = maxExtraSize
    this.maxFetchSize = maxFetchSize
  }

  _canAggregate(requestGroup: ReqGroup, request: Req) {
    return (
      // the fetches overlap, or come close
      request.start <= requestGroup.end + this.maxExtraSize &&
      // aggregating would not result in a fetch that is too big
      request.end - request.start + requestGroup.end - requestGroup.start <
        this.maxFetchSize
    )
  }

  // returns a promise that only resolves when all of the signals in the given
  // array have fired their abort signal
  _allSignalsFired(signals: AbortSignal[]) {
    return new Promise<void>(resolve => {
      let signalsLeft = signals.filter(s => !s.aborted).length
      signals.forEach(signal => {
        signal.addEventListener('abort', () => {
          signalsLeft -= 1
          if (!signalsLeft) {
            resolve()
          }
        })
      })
    }).catch(e => {
      console.error(e)
    })
  }

  // dispatch a request group as a single request and then slice the result
  // back up to satisfy the individual requests
  _dispatch({ url, start, end, requests }: ReqGroup) {
    // if any of the requests have an AbortSignal `signal` in their
    // requestOptions, make our aggregating abortcontroller track it, aborting
    // the request if all of the abort signals that are aggregated here have
    // fired

    const abortWholeRequest = new AbortController()
    const signals = [] as AbortSignal[]
    requests.forEach(({ requestOptions }) => {
      if (requestOptions?.signal) {
        signals.push(requestOptions.signal)
      }
    })
    if (signals.length === requests.length) {
      // may need review
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      this._allSignalsFired(signals).then(() => abortWholeRequest.abort())
    }

    this.fetchCallback(url, start, end - 1, {
      signal: abortWholeRequest.signal,
    }).then(
      response => {
        const data = response.buffer

        requests.forEach(({ start: reqStart, end: reqEnd, resolve }) => {
          resolve({
            headers: response.headers,
            buffer: data.subarray(reqStart - start, reqEnd - start),
          })
        })
      },
      err => {
        requests.forEach(({ reject }) => reject(err))
      },
    )
  }

  _aggregateAndDispatch() {
    Object.entries(this.requestQueues).forEach(([url, requests]) => {
      if (!requests?.length) {
        return
      }

      // we are now going to aggregate the requests in this url's queue into
      // groups of requests that can be dispatched as one
      const requestsToDispatch = [] as Req[]

      // look to see if any of the requests are aborted, and if they are, just
      // reject them now and forget about them
      requests.forEach(request => {
        const { requestOptions, reject } = request
        if (requestOptions?.signal?.aborted) {
          reject(Object.assign(new Error('aborted'), { code: 'ERR_ABORTED' }))
        } else {
          requestsToDispatch.push(request)
        }
      })

      requestsToDispatch.sort((a, b) => a.start - b.start)

      requests.length = 0
      if (!requestsToDispatch.length) {
        return
      }

      let currentRequestGroup: ReqGroup | undefined
      for (const next of requestsToDispatch) {
        if (
          currentRequestGroup &&
          this._canAggregate(currentRequestGroup, next)
        ) {
          // aggregate it into the current group
          currentRequestGroup.requests.push(next)
          currentRequestGroup.end = next.end
        } else {
          // out of range, dispatch the current request group
          if (currentRequestGroup) {
            this._dispatch(currentRequestGroup)
          }
          // and start on a new one
          currentRequestGroup = {
            requests: [next],
            url,
            start: next.start,
            end: next.end,
          }
        }
      }
      if (currentRequestGroup) {
        this._dispatch(currentRequestGroup)
      }
    })
  }

  _enQueue(url: string, request: Req) {
    if (!this.requestQueues[url]) {
      this.requestQueues[url] = []
    }
    this.requestQueues[url].push(request)
  }

  /**
   *
   * @param url
   * @param start 0-based half-open
   * @param end 0-based half-open
   * @param [requestOptions] options passed to the underlying fetch call
   */
  fetch(url: string, start: number, end: number, requestOptions = {}) {
    return new Promise((resolve, reject) => {
      this._enQueue(url, { start, end, resolve, reject, requestOptions })
      if (!this.timeout) {
        this.timeout = setTimeout(() => {
          this.timeout = undefined
          this._aggregateAndDispatch()
        }, this.frequency || 1)
      }
    })
  }
}
