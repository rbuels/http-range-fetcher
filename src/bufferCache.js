const LRU = require('lru-cache')

const CompositeBufferProxyHandler = require('./compositeBufferProxyHandler')
const AggregatingFetcher = require('./aggregatingFetcher')

// TODO: fire events when a remote file is detected as having been changed
/**
 * smart cache that fetches chunks of remote files.
 * caches chunks in an LRU cache, and aggregates upstream fetches
 */
class BufferCache {
  /**
   *
   * @param {object} args
   * @param {function} args.fetch async function with signature (key, start, end) => Promise(Buffer)
   * @param {number} [args.size] size in bytes of cache to keep
   * @param {number} [args.chunkSize] size in bytes of cached chunks
   * @param {number} [args.aggregationTime] time in ms over which to pool requests before dispatching them
   */
  constructor({
    fetch,
    size = 10000000,
    chunkSize = 32768,
    aggregationTime = 100,
  }) {
    if (!fetch) throw new Error('fetch function required')
    this.aggregator = new AggregatingFetcher({
      fetch,
      frequency: aggregationTime,
    })
    this.chunkSize = chunkSize
    this.chunkCache = LRU({ max: Math.floor(size / chunkSize) })
  }

  async get(key, position, length) {
    // calculate the list of chunks involved in this fetch
    const firstChunk = Math.floor(position / this.chunkSize)
    const lastChunk = Math.floor((position + length - 1) / this.chunkSize)

    // fetch them all as necessary
    const fetches = new Array(lastChunk - firstChunk + 1)
    for (let chunk = firstChunk; chunk <= lastChunk; chunk += 1) {
      fetches[chunk - firstChunk] = this._getChunk(key, chunk).then(data => ({
        data,
        chunkNumber: chunk,
      }))
    }

    // return a "composite buffer" that lets the array of chunks be accessed like a flat buffer
    const chunks = await Promise.all(fetches)
    const chunksOffset = position - chunks[0].chunkNumber * this.chunkSize
    return new Proxy(
      chunks,
      new CompositeBufferProxyHandler(chunksOffset, this.chunkSize, length),
    )
  }

  _getChunk(key, chunkNumber) {
    const chunkKey = `${key}/${chunkNumber}`
    const cachedPromise = this.chunkCache.get(chunkKey)
    if (cachedPromise) return cachedPromise

    const freshPromise = this.aggregator.fetch(
      key,
      chunkNumber * this.chunkSize,
      (chunkNumber + 1) * this.chunkSize,
    )
    this.chunkCache.set(chunkKey, freshPromise)
    return freshPromise
  }
}

module.exports = BufferCache
