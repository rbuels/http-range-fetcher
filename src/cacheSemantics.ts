//@ts-nocheck
export function parseCacheControl(field) {
  if (typeof field !== 'string') {
    return {}
  }

  const parsed = {}
  const invalid = field.toLowerCase().replace(
    // eslint-disable-next-line no-control-regex,no-useless-escape
    /(?:^|(?:\s*,\s*))([^\x00-\x20\(\)<>@\,;\:\\"\/\[\]\?\=\{\}\x7F]+)(?:\=(?:([^\x00-\x20\(\)<>@\,;\:\\"\/\[\]\?\=\{\}\x7F]+)|(?:\"((?:[^"\\]|\\.)*)\")))?/g,
    (match, fieldName, three, four) => {
      const value = three || four
      parsed[fieldName] = value ? value.toLowerCase() : true
      return ''
    },
  )

  if (invalid) {
    return {}
  }

  // parse any things that seem to be numbers
  Object.keys(parsed).forEach(key => {
    if (/^[\d]+$/.test(parsed[key])) {
      try {
        const num = parseInt(parsed[key], 10)
        if (!Number.isNaN(num)) {
          parsed[key] = num
        }
      } catch (e) {
        /* ignore */
      }
    }
  })

  return parsed
}

export class CacheSemantics {
  constructor({ minimumTTL }) {
    this.minimumTTL = minimumTTL
  }

  calculateChunkExpirationDate(chunkResponse) {
    const { headers = {}, requestDate, responseDate } = chunkResponse
    let baselineDate = responseDate || requestDate
    if (!baselineDate) {
      if (!headers.date) {
        return undefined
      }
      baselineDate = new Date(headers.date)
    }

    const basePlus = ttl => new Date(baselineDate.getTime() + ttl)

    // results that are not really cacheable expire after the minimum time to live
    if (/\bno-cache\b/.test(headers.pragma)) {
      return basePlus(this.minimumTTL)
    }

    const cacheControl = parseCacheControl(headers['cache-control'])
    if (
      cacheControl['no-cache'] ||
      cacheControl['no-store'] ||
      cacheControl['must-revalidate']
    ) {
      return basePlus(this.minimumTTL)
    }

    if (cacheControl['max-age'] !== undefined) {
      const ttl = cacheControl['max-age'] * 1000 // max-age is in seconds
      return basePlus(Math.max(ttl, this.minimumTTL))
    } else if (this._coerceToDate(headers.expires)) {
      return this._coerceToDate(headers.expires)
    } else if (this._coerceToDate(headers['last-modified'])) {
      const lastModified = this._coerceToDate(headers['last-modified'])
      const ttl = (baselineDate.getTime() - lastModified.getTime()) / 10
      return basePlus(ttl)
    }

    // otherwise, we just cache forever
    return undefined
  }

  _coerceToDate(thing) {
    if (thing) {
      if (thing instanceof Date) {
        return thing
      }
      if (typeof thing === 'string' || typeof thing === 'number') {
        return new Date(thing)
      }
    }
    return undefined
  }

  /**
   * check whether a cached chunk response is still valid and can be used
   * @param {object} chunkResponse
   * @returns {boolean}
   */
  cachedChunkIsValid(chunkResponse) {
    const expiration = this.calculateChunkExpirationDate(chunkResponse)
    return !expiration || new Date() <= expiration
  }

  /**
   * check whether the response for this chunk fetch can be cached
   * @param {object} chunkResponse
   * @returns {boolean}
   */
  chunkIsCacheable() {
    // right now, we are caching everything, we just give it a very short
    // time to live if it's not supposed to be cached
    return true
  }
}
