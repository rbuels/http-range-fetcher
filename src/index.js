const HttpRangeCache = require('./httpRangeCache')

module.exports = {
  HttpRangeCache,
}
