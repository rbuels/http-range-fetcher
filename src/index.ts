//@ts-nocheck
import HttpRangeFetcher from './httpRangeFetcher'

export { HttpRangeFetcher }
