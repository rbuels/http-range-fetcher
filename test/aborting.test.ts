//@ts-nocheck
import { it, expect } from 'vitest'
import { HttpRangeFetcher } from '../src/index'

const timeout = ms => new Promise(res => setTimeout(res, ms))

it(`can abort a fetch 1`, async () => {
  expect.assertions(2)
  const ab = new AbortController()
  const calls = []
  async function fetch(url, start, end, options) {
    calls.push([url, start, end, options])
    if (options.signal.aborted) {
      throw Object.assign(new Error('aborted'), { code: 'ERR_ABORTED' })
    }
    return {
      headers: {},
      responseDate: new Date(),
      buffer: Buffer.from([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
    }
  }
  const cache = new HttpRangeFetcher({ fetch, aggregationTime: 100 })
  const get = cache.getRange('http://foo.com/', 0, 10, { signal: ab.signal })
  ab.abort()
  await expect(get).rejects.toThrow(/aborted/)
  expect(calls).toMatchSnapshot()
})

it(`can abort a fetch 2`, async () => {
  expect.assertions(2)
  const ab = new AbortController()
  const calls = []
  async function fetch(url, start, end, options) {
    calls.push([url, start, end, options])
    // await new Promise(res => process.nextTick(res))
    if (options.signal.aborted) {
      throw Object.assign(new Error('aborted'), { code: 'ERR_ABORTED' })
    }
    await timeout(100)
    return {
      headers: {},
      responseDate: new Date(),
      buffer: Buffer.from([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
    }
  }
  const cache = new HttpRangeFetcher({ fetch, aggregationTime: 100 })
  const get = cache.getRange('http://foo.com/', 0, 10, { signal: ab.signal })
  await timeout(50)
  ab.abort()
  await expect(get).rejects.toThrow(/aborted/)
  expect(calls).toMatchSnapshot()
})

it(`can abort a fetch 3`, async () => {
  expect.assertions(3)
  const ab = new AbortController()
  const calls = []
  let abortCount = 0
  async function fetch(url, start, end, options) {
    calls.push([url, start, end, options])
    await timeout(50)
    if (options.signal.aborted) {
      abortCount += 1
      throw Object.assign(new Error('aborted'), { code: 'ERR_ABORTED' })
    }
    return {
      headers: {},
      responseDate: new Date(),
      buffer: Buffer.from([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
    }
  }
  const cache = new HttpRangeFetcher({ fetch, aggregationTime: 1 })
  const get = cache.getRange('http://foo.com/', 0, 10, { signal: ab.signal })
  await timeout(20)
  ab.abort()
  await expect(get).rejects.toThrow(/aborted/)
  expect(calls).toMatchSnapshot()
  expect(abortCount).toBe(1)
})
