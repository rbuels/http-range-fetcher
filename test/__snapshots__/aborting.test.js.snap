// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can abort a fetch 1 1`] = `Array []`;

exports[`can abort a fetch 2 1`] = `Array []`;

exports[`can abort a fetch 3 1`] = `
Array [
  Array [
    "http://foo.com/",
    0,
    32767,
    Object {
      "signal": AbortSignal {},
    },
  ],
]
`;
